### ISSUE TITLE
EOF when loading the same chart concurrently

### ISSUE THREAD

**Issue by aaronmell** [helm/helm#6140]
We have a bunch of microservices that use the same helm chart. We are using the terraform provider for helm, which deploys all of our microservices concurrently. We randomly see EOF errors returning from the provider. I've traced the issue into the LoadFile function inside the chartutil by adding logging to the terraform provider.

Failed to load chart from path /var/lib/jenkins/.helm/cache/archive/microservice-0.8.4.tgz, EOF

Output of `helm version`:
Client: &version.Version{SemVer:"v2.14.2", GitCommit:"a8b13cc5ab6a7dbef0a58f5061bcc7c0c61598e7", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.14.2", GitCommit:"a8b13cc5ab6a7dbef0a58f5061bcc7c0c61598e7", GitTreeState:"clean"}

Output of `kubectl version`:
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", BuildDate:"2018-12-13T19:44:19Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.6-eks-d69f1b", GitCommit:"d69f1bf3669bf00b7f4a758e978e0e7a1e3a68f7", GitTreeState:"clean", BuildDate:"2019-02-28T20:26:10Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
EKS

**Comment by karuppiah7890** [comment]
I have actually run into this too in our pipeline. Never thought of submitting the issue. Triggering pipeline solved the problem. There was one more regarding some "pipe broken" issue. I assumed it was some port forward issue or something and again pipeline retrigger solved it. 

I think I should start checking all such issues and submit it.

Regarding this issue, I'll try to find a proper way to reproduce this and try to raise a PR to fix this :)

**Comment by karuppiah7890** [comment]
I have actually run into this too in our pipeline. Never thought of submitting the issue. Triggering pipeline solved the problem. There was one more regarding some "pipe broken" issue. I assumed it was some port forward issue or something and again pipeline retrigger solved it. 

I think I should start checking all such issues and submit it.

Regarding this issue, I'll try to find a proper way to reproduce this and try to raise a PR to fix this :)

**Comment by karuppiah7890** [comment]
@aaronmell After quite some thinking and checking cases when this error occurred, you are right that this occurs only when loading same chart concurrently. We have faced similar issues too. On digging further, I found that the chart is downloaded if it's not present. Now when the download happens, if another helm command is run concurrently, and if it finds that the helm chart tar ball is present , it just tries to load the chart tar ball. Now the issue comes when the tar ball is not completely downloaded by the first command. In which case the load fails. This is a classic concurrency problem. We need to see our options and choose one to solve it. And I think this can occur only for charts that are being downloaded to the cache directory. Local chart repos, and tar balls wouldn't have the problem - as all helm commands will only read - writing is already done. So issue kicks in only when downloading and loading chart happens concurrently in two different commands, and download is not complete yet. So it's writing and reading concurrently which causes issues.
 
I think the current facts are - 
1. Download happens only once, for a given chart and version, and then the chart can be read multiple times - but only after download finishes. Or else the read will be a dirty read leading to errors.

The solutions need to be around the above. Some thoughts : 

1. Since it's a download, it's coming from a repo which we have added locally and for which index is present. To check if a download is over, we can use the sha 256 sum in the index file and then compare it with the sha 256 sum of the file. The other commands can wait till the download is over by checking the shasum, and download doesn't take much time as charts are small, so the time for which other commands wait would be less
2. Point 1 can be a problematic solution where - if for some reason, the download was stopped in between, and just halted. In which case, all the other helm commands would just keep waiting for the download to be over, which is probably infinite. So, we need to handle this scenario by somehow letting the other helm commands resume existing download or do fresh download and resume it's work (installation, upgradation etc)

What do you think ? I'll also go check what other theories say about such scenarios. For example lock mechanism or something that can help?

cc @bacongobbler

**Comment by karuppiah7890** [comment]
I also wrote a test to validate this in this file

https://github.com/helm/helm/blob/master/pkg/chartutil/load_test.go

```
t.Run("should load same tar ball in concurrent calls", func(t *testing.T) {
	chart := "testdata/frobnitz-1.2.3.tgz"
	concurrentDownloadingChart := "testdata/concurrent-downloading-chart.tgz"

	fs := afero.NewOsFs()

	data, err := afero.ReadFile(fs, chart)
	assert.NoError(t, err)

	// simulating situation like - download in progress (50% downloaded)
	err = afero.WriteFile(fs, concurrentDownloadingChart, data[:len(data)/2], 0666)
	assert.NoError(t, err)

	// loading chart when only 50% of chart is downloaded
	c, err := Load(concurrentDownloadingChart)
	if err != nil {
		t.Fatalf("Failed to load testdata: %s", err)
		return
	}
	verifyFrobnitz(t, c)
	verifyChart(t, c)
	verifyRequirements(t, c)
})
```

It's only a simulation of the above speculation. If I'm missing anything, or you think there's a different issue here, let me know 👍

**Comment by aaronmell** [comment]
That seems reasonable.

**Comment by AlbertoPeon** [comment]
Hi @aaronmell @karuppiah7890 !

Thanks a lot for the investigation - do you guys know if this issue is planned to be fixed in a future release of Helm 2? I am also considering managing Helm charts with Terraform but this is issue seems to be a blocker to run this in production.

Did you find a workaround to make it work reliably @aaronmell ?

**Comment by karuppiah7890** [comment]
@AlbertoPeon I was expecting #4106 to help this. Will check it out. Would you be willing to raise a PR similar to #4106 ?

Hey @bacongobbler , do you think someone can start working on a PR similar to #4106 ? As that PR has no responses

**Comment by mauriciopoppe** [comment]
@AlbertoPeon @aaronmell 

As a workaround you can download the chart manually to a temp location and then install the chart from the filesystem, in the following example a new temp directory is created for every new chart version

```hcl
resource "random_string" "random" {
  length = 16
  special = false
  keepers = {
    version = "${var.chart_version}"
  }
}

locals {
  chart_location = "/tmp/${random_string.random.result}"
}

resource "null_resource" "helm_pull" {
  triggers = {
    version = "${var.chart_version}"
  }

  provisioner "local-exec" {
    command = "mkdir -p ${local.chart_location}"
  }

  provisioner "local-exec" {
    command = "helm repo add myrepo ${var.myrepo_url}"
  }

  provisioner "local-exec" {
    command = "helm repo update"
  }

  provisioner "local-exec" {
    command = "helm fetch myrepo/mychart --version ${var.chart_version} --destination ${local.chart_location} --untar"
  }
}

resource "helm_release" "mychart" {
  ...
  chart = "${local.chart_location}/mychart"
  ...
}
```

**Comment by yurrriq** [comment]
Seems like a duplicate of #7600, which was resolved with #7720 for Helm >=v3.2.0.
